File: robotics_planning/grid.py
"""Occupancy-grid primitives used by the planning experiments."""

from __future__ import annotations

from dataclasses import dataclass, field
import textwrap
from typing import Iterable, Mapping

Cell = tuple[int, int]


@dataclass(frozen=True)
class GridMap:
    """A small 2-D occupancy grid.

    Coordinates are ``(x, y)``. ``x`` increases to the right and ``y``
    increases downward when the grid is rendered as ASCII text.

    ``cell_costs`` stores positive costs for entering selected free cells.
    Cells not present in the mapping have cost 1.0.
    """

    width: int
    height: int
    obstacles: frozenset[Cell] = field(default_factory=frozenset)
    cell_costs: Mapping[Cell, float] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if self.width <= 0 or self.height <= 0:
            raise ValueError("Grid dimensions must be positive")
        for cell in self.obstacles:
            if not self.in_bounds(cell):
                raise ValueError(f"Obstacle outside grid: {cell}")
        for cell, cost in self.cell_costs.items():
            if not self.in_bounds(cell):
                raise ValueError(f"Cost cell outside grid: {cell}")
            if cost <= 0:
                raise ValueError("Cell costs must be positive")

    def in_bounds(self, cell: Cell) -> bool:
        x, y = cell
        return 0 <= x < self.width and 0 <= y < self.height

    def is_free(self, cell: Cell) -> bool:
        return self.in_bounds(cell) and cell not in self.obstacles

    def cost_to_enter(self, cell: Cell) -> float:
        if not self.is_free(cell):
            raise ValueError(f"Cannot enter occupied or out-of-bounds cell: {cell}")
        return float(self.cell_costs.get(cell, 1.0))

    def neighbors(self, cell: Cell, diagonal: bool = False) -> tuple[Cell, ...]:
        """Return free neighboring cells in deterministic order."""

        x, y = cell
        directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        if diagonal:
            directions += [(1, 1), (-1, 1), (-1, -1), (1, -1)]
        return tuple(
            candidate
            for dx, dy in directions
            if self.is_free(candidate := (x + dx, y + dy))
        )

    def inflated(self, radius_cells: int) -> "GridMap":
        """Return a copy with obstacles expanded by a square cell radius."""

        if radius_cells < 0:
            raise ValueError("Inflation radius cannot be negative")
        inflated_obstacles = set(self.obstacles)
        for obstacle in self.obstacles:
            ox, oy = obstacle
            for dx in range(-radius_cells, radius_cells + 1):
                for dy in range(-radius_cells, radius_cells + 1):
                    candidate = (ox + dx, oy + dy)
                    if self.in_bounds(candidate):
                        inflated_obstacles.add(candidate)
        return GridMap(
            width=self.width,
            height=self.height,
            obstacles=frozenset(inflated_obstacles),
            cell_costs=self.cell_costs,
        )

    def render(
        self,
        path: Iterable[Cell] = (),
        explored: Iterable[Cell] = (),
        start: Cell | None = None,
        goal: Cell | None = None,
    ) -> str:
        """Render the grid, explored cells, and final path as ASCII."""

        path_set = set(path)
        explored_set = set(explored)
        rows: list[str] = []
        for y in range(self.height):
            row: list[str] = []
            for x in range(self.width):
                cell = (x, y)
                if cell in self.obstacles:
                    symbol = "#"
                elif cell == start:
                    symbol = "S"
                elif cell == goal:
                    symbol = "G"
                elif cell in path_set:
                    symbol = "*"
                elif cell in explored_set:
                    symbol = "+"
                else:
                    symbol = "."
                row.append(symbol)
            rows.append("".join(row))
        return "\n".join(rows)


def parse_ascii_map(lines: str | Iterable[str]) -> tuple[GridMap, Cell, Cell]:
    """Parse an ASCII map and return ``(grid, start, goal)``."""

    raw_lines = (
        textwrap.dedent(lines).splitlines()
        if isinstance(lines, str)
        else list(lines)
    )
    cleaned = [line.rstrip("\n") for line in raw_lines if line.strip()]
    if not cleaned:
        raise ValueError("Map cannot be empty")

    width = len(cleaned[0])
    if width == 0 or any(len(line) != width for line in cleaned):
        raise ValueError("Map must be rectangular")

    obstacles: set[Cell] = set()
    start: Cell | None = None
    goal: Cell | None = None
    for y, line in enumerate(cleaned):
        for x, symbol in enumerate(line):
            cell = (x, y)
            if symbol == "#":
                obstacles.add(cell)
            elif symbol == "S":
                if start is not None:
                    raise ValueError("Map must contain exactly one start")
                start = cell
            elif symbol == "G":
                if goal is not None:
                    raise ValueError("Map must contain exactly one goal")
                goal = cell
            elif symbol != ".":
                raise ValueError(f"Unsupported map symbol: {symbol!r}")

    if start is None or goal is None:
        raise ValueError("Map must contain one start and one goal")
    return GridMap(width, len(cleaned), frozenset(obstacles)), start, goal

File: cpp/src/grid.cpp
#include "robotics_planning/grid.hpp"

#include <algorithm>
#include <cmath>
#include <sstream>
#include <stdexcept>
#include <utility>

namespace robotics_planning {

GridMap::GridMap(
    int width,
    int height,
    std::vector<std::uint8_t> occupied,
    std::vector<double> cell_costs
) : width_(width), height_(height) {
    if (width_ <= 0 || height_ <= 0) {
        throw std::invalid_argument("Grid dimensions must be positive");
    }

    const auto expected_size = static_cast<std::size_t>(width_ * height_);
    if (occupied.empty()) {
        occupied.assign(expected_size, 0U);
    } else if (occupied.size() != expected_size) {
        throw std::invalid_argument("Occupancy array has the wrong size");
    }

    if (cell_costs.empty()) {
        cell_costs.assign(expected_size, 1.0);
    } else if (cell_costs.size() != expected_size) {
        throw std::invalid_argument("Cell-cost array has the wrong size");
    }

    for (const double cost : cell_costs) {
        if (cost <= 0.0 || !std::isfinite(cost)) {
            throw std::invalid_argument("Cell costs must be finite and positive");
        }
    }

    occupied_ = std::move(occupied);
    cell_costs_ = std::move(cell_costs);
}

bool GridMap::in_bounds(int x, int y) const noexcept {
    return x >= 0 && x < width_ && y >= 0 && y < height_;
}

bool GridMap::in_bounds(Cell cell) const noexcept {
    return cell >= 0 && cell < cell_count();
}

bool GridMap::is_free(Cell cell) const noexcept {
    return in_bounds(cell) && occupied_[static_cast<std::size_t>(cell)] == 0U;
}

double GridMap::cost_to_enter(Cell cell) const {
    if (!is_free(cell)) {
        throw std::invalid_argument("Cannot enter an occupied or invalid cell");
    }
    return cell_costs_[static_cast<std::size_t>(cell)];
}

std::array<Cell, 4> GridMap::neighbors4(Cell current) const noexcept {
    if (!in_bounds(current)) {
        return {-1, -1, -1, -1};
    }

    const int x = x_of(current);
    const int y = y_of(current);
    const std::array<std::pair<int, int>, 4> directions = {
        std::pair<int, int>{1, 0},
        std::pair<int, int>{0, 1},
        std::pair<int, int>{-1, 0},
        std::pair<int, int>{0, -1},
    };

    std::array<Cell, 4> result{};
    for (std::size_t index = 0; index < directions.size(); ++index) {
        const int neighbor_x = x + directions[index].first;
        const int neighbor_y = y + directions[index].second;
        const Cell candidate = in_bounds(neighbor_x, neighbor_y)
            ? cell(neighbor_x, neighbor_y)
            : -1;
        result[index] = candidate >= 0 && is_free(candidate) ? candidate : -1;
    }
    return result;
}

std::string GridMap::render(
    const std::vector<Cell>& path,
    const std::vector<Cell>& expanded,
    Cell start,
    Cell goal
) const {
    std::vector<char> symbols(static_cast<std::size_t>(cell_count()), '.');
    for (Cell cell_index = 0; cell_index < cell_count(); ++cell_index) {
        if (!is_free(cell_index)) {
            symbols[static_cast<std::size_t>(cell_index)] = '#';
        }
    }

    for (const Cell cell_index : expanded) {
        if (is_free(cell_index)) {
            symbols[static_cast<std::size_t>(cell_index)] = '+';
        }
    }
    for (const Cell cell_index : path) {
        if (is_free(cell_index)) {
            symbols[static_cast<std::size_t>(cell_index)] = '*';
        }
    }
    if (is_free(start)) {
        symbols[static_cast<std::size_t>(start)] = 'S';
    }
    if (is_free(goal)) {
        symbols[static_cast<std::size_t>(goal)] = 'G';
    }

    std::ostringstream output;
    for (int y = 0; y < height_; ++y) {
        for (int x = 0; x < width_; ++x) {
            output << symbols[static_cast<std::size_t>(cell(x, y))];
        }
        if (y + 1 < height_) {
            output << '\n';
        }
    }
    return output.str();
}

ParsedAsciiMap parse_ascii_map(const std::vector<std::string>& lines) {
    if (lines.empty()) {
        throw std::invalid_argument("Map cannot be empty");
    }

    const int height = static_cast<int>(lines.size());
    const int width = static_cast<int>(lines.front().size());
    if (width == 0) {
        throw std::invalid_argument("Map rows cannot be empty");
    }

    std::vector<std::uint8_t> occupied(
        static_cast<std::size_t>(width * height), 0U
    );
    Cell start = -1;
    Cell goal = -1;

    for (int y = 0; y < height; ++y) {
        if (static_cast<int>(lines[static_cast<std::size_t>(y)].size()) != width) {
            throw std::invalid_argument("Map must be rectangular");
        }

        for (int x = 0; x < width; ++x) {
            const char symbol = lines[static_cast<std::size_t>(y)][static_cast<std::size_t>(x)];
            const Cell current = y * width + x;
            if (symbol == '#') {
                occupied[static_cast<std::size_t>(current)] = 1U;
            } else if (symbol == 'S') {
                if (start != -1) {
                    throw std::invalid_argument("Map must contain one start");
                }
                start = current;
            } else if (symbol == 'G') {
                if (goal != -1) {
                    throw std::invalid_argument("Map must contain one goal");
                }
                goal = current;
            } else if (symbol != '.') {
                throw std::invalid_argument("Unsupported map symbol");
            }
        }
    }

    if (start == -1 || goal == -1) {
        throw std::invalid_argument("Map must contain one start and one goal");
    }

    return {GridMap(width, height, std::move(occupied)), start, goal};
}

}  // namespace robotics_planning
